proposals/0000-dependent-type-design.rst

```diff
@@ -962,6 +962,63 @@ will treat these two very differently:
 Even once we reach the Glorious Day, nothing forces us to unify ``t1 -> t2``
 with ``foreach (_ :: t1) -> t2``, and we may decide not to.
 
+Learnability of Haskell
+^^^^^^^^^^^^^^^^^^^^^^^
+
+A cross-cutting concern in the design of depdendent types in Haskell is whether
+they will make learning Haskell more difficult for newcomers to the language, or
+less usable for long-time Haskellers who prefer to avoid dependent types.
+
+We thus set forth this principle:
+
+**The Opt-In Principle (OIP):** Users who do not opt into dependent types will
+not be affected by them.
+
+By "opt into", we mean that users would have to enable ``-XDependentTypes`` or
+import a module that exposes functions with depenently-typed interfaces. These
+modules would not be standard modules that are routinely imported today, such
+as ``Data.List`` or ``Prelude``.
+
+Like all principles, we cannot promise that there will not be exceptions, but
+any exceptions made would be made wary of the OIP. An example from the history
+of adding fancy types to GHC is around the type of ``($)``, which became levity-polymorphic
+for GHC 8.0, despite the fact that ``($)`` is exported in ``Prelude``. The solution
+is to have ``-fprint-explicit-runtime-reps``, off by default, that allows users
+to see the full type of ``($)``. Without ``-fprint-explicit-runtime-reps``, users see
+a simplified type ``(a -> b) -> a -> b``, as they might expect. Extrapolating to
+dependent types, if a function in ``Prelude`` were to get a dependent type, we
+would design the new type to be backward compatible and to be hidden behind a flag
+similar to ``-fprint-explicit-runtime-reps``. It is our hope that efforts toward
+better IDE support for Haskell will make such designs easier to contemplate, where
+a user might, say, click on a confusing aspect of an error message to get more detail.
+
+Another important aspect of the OIP is in error messages. Suppose my second day of
+Haskell includes ::
+
+  x :: Just Int
+  x = Nothing
+
+I would then see a suggestion ``A data constructor of that name is in scope;
+did you mean DataKinds?``. This suggestion is unhelpful on the second day of
+Haskelling: much better would be for GHC to suggest that the user meant ``Maybe``
+instead of ``Just``. There is an inherent tension between keeping users in
+a language subset that is more understandable and discoverability. One possibility
+would be for an error message to label how "advanced" a language extension is
+as it is being suggested, or to link to a page with more information and guidance.
+If our second-day-of-Haskell user were told
+``Did you mean to enable advanced extension DataKinds?`` perhaps they would
+seek other fixes to their program before enabling ``-XDataKinds``.
```

Comment:
What we have today is already pretty challenging. The example proposed here is not going to be a significant improvement for newbs. Elm gets this right, and the Haskell community could learn from that experience.